File: Lab5/Prog2_Lab5/Prog2_Lab5_Lib_C++/Figure.cpp
#include "pch.h"
#include "Figure.h"

// Метод отримання даних (точок) фігури
vector<vector<int>> Figure::Get() {
	return points;
};

// Перевантажений метод отримання даних фігури, повертає одне значення матриці точок
// Використовується в функції виведення значень на екран, в основній програмі
int Figure::Get(int i, int j) {
	return points[i][j];
};

// Ініціалізатор для фігур
// Задає розмір вектора
void Figure::_Init(int n) {
	points.assign(n, { 0, 0 });
}

// Пустий конструктор для використання підкласами з власними конструкторами
Figure::Figure() {
}

// Конструктор з памаметром
Figure::Figure(int input[4][2]) {
	_Init(4);

	for (int i = 0; i < 4; i++) {
		for (int j = 0; j < 2; j++) {
			points[i][j] = input[i][j];
		}
	}
}

// Метод обчислення відстанні між точками фігури
float Figure::Length(int i1, int i2) {
	if (i1 == -1) { i1 = 3; }
	vector<int> a(2);
	vector<int> b(2);

	for (int i = 0; i < 2; i++) {
		a[i] = points[i1][i];
		b[i] = points[i2][i];
	}

	return sqrtf(pow((double(a[0]) - double(b[0])), 2) + pow((double(a[1]) - double(b[1])), 2));
};

// Ініціалізатор для квадрату, задає розмір вектора
// 4 точки, 2 виміри
void Square::_Init() {
	points.assign(4, { 0, 0 });
}

// Конструктор з памаметром
Square::Square(int input[4][2]) {
	_Init();

	for (int i = 0; i < 4; i++) {
		for (int j = 0; j < 2; j++) {
			points[i][j] = input[i][j];
		}	
	}
}

// Метод обчислення периметра
float Square::Perimeter() {
	float peri = 0;
	for (int i = 0; i < 3; i++)
	{
		peri += Length(i - 1, i);
	}
	return peri;
};

// Метод обчислення площі
float Square::Area() {
	return Length(0, 1) * Length(1, 2);
};

File: Lab5/Prog2_Lab5/Prog2_Lab5_1_Cs_Lib/Figure.cs
using System;

namespace Prog2_Lab5_1_Cs_Lib
{
    public class Figure
    {
        protected int[,] points;

        public int[,] Get()
        {
            return points;
        }
        public int Get(int i1, int i2)
        {
            return points[i1, i2];
        }

        // Конструктор з памаметром
        public Figure(int[,] input)
        {
            points = input;
        }

        // Метод для розрахунку відстані між вершинами
        public float Length(int i1, int i2)
        {
            if (i1 == -1) { i1 = 3; }
            int[] a = { points[i1, 0], points[i1, 1] };
            int[] b = { points[i2, 0], points[i2, 1] };
            return Convert.ToSingle(Math.Sqrt(Math.Pow((a[0] - b[0]), 2) + Math.Pow((a[1] - b[1]), 2)));
        }

        // Пустий конструктор для використання підкласами з власними конструкторами
        // protected залишає доступ похідним классам без повного відкриття команди
        protected Figure()
        {
        }
    }

    public class Square : Figure
    {
        // Конструктор з памаметром
        public Square(int[,] input)
        {
            points = new int[4, 2];
            for (int i = 0; i < 4; i++)
            {
                for (int j = 0; j < 2; j++)
                {
                    points[i, j] = input[i, j];
                }
            }
        }

        // Метод для розрахунку периметра квадрату
        // Використовує метод визначення відстані з базового класу
        public float Perimeter()
        {
            float peri = 0;
            for (int i = 0; i < 4; i++)
            {
                peri += Length(i - 1, i);
            }
            return peri;
        }

        // Метод для розрахунку площі квадрату
        // Використовує метод визначення відстані з базового класу
        public float Area()
        {
            return (Length(0, 1) * Length(1, 2));
        }
    }
}
